# Pandas en Python

Se importa la librería **pandas**

In [3]:
import pandas as pd

## DataFrame

### Crear un DataFrame

#### Nota
- pd.DataFrame() es una función de pandas que crea una tabla (como Excel o una tabla SQL).
- Cada clave del diccionario se vuelve una columna.
- Cada lista dentro del diccionario se vuelve una fila de datos.

#### 1. Diccionario de listas(columna por columna)

In [4]:
football_dict = {
    "player": ["Lionel Messi", "Cristiano Ronaldo"],
    "year": [2016, 2016],
    "goals": [37, 25],
}
football_stats = pd.DataFrame(football_dict)
print(football_stats)

              player  year  goals
0       Lionel Messi  2016     37
1  Cristiano Ronaldo  2016     25


#### 2. Lista de diccionarios(fila por fila)

In [5]:
football_list = [
    {"player": "Lionel Messi", "year": 2016, "goals": 37},
    {"player": "Cristiano Ronaldo", "year": 2016, "goals": 25},
]
football_stats = pd.DataFrame(football_list)
print(football_stats)

              player  year  goals
0       Lionel Messi  2016     37
1  Cristiano Ronaldo  2016     25


## Importar datos

### Base de datos

#### Se usara una base de datos de las 5 ligas top de Europa entre los años 2014 y 2020.

### Importar datos de un archivo csv

In [7]:
df = pd.read_csv('Fullmetadata.csv', index_col= 'player_id')
df

player_name  games  time  goals        xG  assists  \
player_id                                                               
8865             Ollie Watkins      9   810      6  6.108615        1   
675              Jack Grealish      9   810      5  3.136721        5   
592               Ross Barkley      6   454      2  1.515623        1   
1024              Tyrone Mings      9   810      2  1.362349        1   
7726          Ezri Konsa Ngoyo      9   810      2  0.972794        0   
...                        ...    ...   ...    ...       ...      ...   
4267          Florian Hartherz     10   775      0  0.008183        2   
4268       Christian Strohdiek     22  1608      0  0.192864        0   
4311                Idir Ouali      6   296      0  0.608952        0   
4334              Mirnes Pepic      2    20      0  0.000000        0   
4363            Thomas Bertels      1     3      0  0.000000        0   

                 xA  shots  key_passes  yellow_cards  red_cards position  \
player_id                                                                  
8865       1.294178     22          12             0          0        F   
675        3.884479     26          26             2          0      F M   
592        1.767355     17          15             0          0        M   
1024       0.068672      7           2             2          0        D   
7726       0.000000      7           0             0          0        D   
...             ...    ...         ...           ...        ...      ...   
4267       0.341444      1           5             1          1      D S   
4268       0.095853      6           3             2          0      D S   
4311       0.054503      4           1             1          0      M S   
4334       0.000000      0           0             0          0        S   
4363       0.000000      0           0             0          0        S   

             team_name  npg      npxG   xGChain  xGBuildup  year  
player_id                                                         
8865       Aston Villa    5  4.586278  6.570879   1.832243  2020  
675        Aston Villa    5  3.136721  7.380403   2.577331  2020  
592        Aston Villa    2  1.515623  3.546400   0.667209  2020  
1024       Aston Villa    2  1.362349  1.677874   1.661455  2020  
7726       Aston Villa    2  0.972794  0.627418   0.627418  2020  
...                ...  ...       ...       ...        ...   ...  
4267         Paderborn    0  0.008183  1.099364   0.908752  2014  
4268         Paderborn    0  0.192864  1.002443   0.892720  2014  
4311         Paderborn    0  0.608952  0.983527   0.374576  2014  
4334         Paderborn    0  0.000000  0.000000   0.000000  2014  
4363         Paderborn    0  0.000000  0.000000   0.000000  2014  

[18633 rows x 18 columns]

## Inspeccionar un DataFrame

- `df.head()` → retorna las 5 primeras filas. También le puedes pasar un entero como parámetro y te devolverá esa cantidad de primeras filas.
- `df.tail()` → retorna las últimas 5 filas del dataset.
- `df.sample()` → es parecido a los 2 anteriores, pero este tomará muestras al azar del DataFrame.
- `df.shape` → retorna las filas y columnas que tiene el DataFrame.
- `df.size` → multiplica las filas y columnas y te da el total de datos del DataFrame.
- `df.info()` → este es genial, te da la cuenta de valores no nulos, el tipo de dato de cada columna (recuerda que solo puede haber un único tipo de dato por columna) y el uso de memoria. Cuando estés procesando los datos, será un gran aliado.
- `df.describe()` → te ayudará mucho con las primeras impresiones de los datos. Calcula algunas estadísticas descriptivas para cada

In [8]:
print(df.sample(3))

                player_name  games  time  goals         xG  assists        xA  \
player_id                                                                       
504            Aaron Ramsey     23  1212      1   3.505756        4  4.432391   
2048       Antonio Nocerino      2    66      0   0.000000        0  0.083201   
2410                 Aduriz     31  2586     18  16.696459        4  4.446756   

           shots  key_passes  yellow_cards  red_cards position      team_name  \
player_id                                                                       
504           42          23             3          0      M S        Arsenal   
2048           0           1             0          0      M S       AC Milan   
2410          88          17            12          0      F S  Athletic Club   

           npg       npxG    xGChain  xGBuildup  year  
player_id                                              
504          1   3.505756  11.132135   5.294314  2016  
2048         0   0.0

In [9]:
print(df.shape)

(18633, 18)


In [10]:
print(df.size)

335394


In [11]:
print(df.info())

<class 'pandas.DataFrame'>
Index: 18633 entries, 8865 to 4363
Data columns (total 18 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   player_name   18633 non-null  str    
 1   games         18633 non-null  int64  
 2   time          18633 non-null  int64  
 3   goals         18633 non-null  int64  
 4   xG            18633 non-null  float64
 5   assists       18633 non-null  int64  
 6   xA            18633 non-null  float64
 7   shots         18633 non-null  int64  
 8   key_passes    18633 non-null  int64  
 9   yellow_cards  18633 non-null  int64  
 10  red_cards     18633 non-null  int64  
 11  position      18633 non-null  str    
 12  team_name     18633 non-null  str    
 13  npg           18633 non-null  int64  
 14  npxG          18633 non-null  float64
 15  xGChain       18633 non-null  float64
 16  xGBuildup     18633 non-null  float64
 17  year          18633 non-null  int64  
dtypes: float64(5), int64(10), str(3)
memory 

In [12]:
print(df.describe())

              games          time         goals            xG       assists  \
count  18633.000000  18633.000000  18633.000000  18633.000000  18633.000000   
mean      17.072291   1207.483282      1.624054      1.660439      1.138088   
std       11.375527    975.753086      3.252418      2.974699      1.936238   
min        1.000000      1.000000      0.000000      0.000000      0.000000   
25%        7.000000    326.000000      0.000000      0.060762      0.000000   
50%       16.000000    976.000000      0.000000      0.560141      0.000000   
75%       27.000000   2004.000000      2.000000      1.900353      2.000000   
max       38.000000   3420.000000     48.000000     39.308761     20.000000   

                 xA         shots    key_passes  yellow_cards     red_cards  \
count  18633.000000  18633.000000  18633.000000  18633.000000  18633.000000   
mean       1.165425     15.323888     11.354801      2.453979      0.127569   
std        1.735366     20.934645     15.065863    

## Ordenar los datos

 `df.sort_values()` Esta función en Pandas se usa para ordenar un DataFrame según los valores de una o más columnas. De forma predeterminada, ordena en orden ascendente y devuelve un nuevo DataFrame, dejando el original sin cambios a menos que se especifique lo contrario.

In [13]:
df.sort_values('player_name', ascending=False)

player_name  games  time  goals        xG  assists        xA  \
player_id                                                                     
250           Ørjan Nyland      6   540      0  0.000000        0  0.000000   
250           Ørjan Nyland     12  1021      0  0.000000        0  0.000000   
250           Ørjan Nyland      7   536      0  0.000000        0  0.000000   
2646        Özkan Yildirim      1    60      0  0.000000        0  0.000000   
2646        Özkan Yildirim      1    13      0  0.083136        0  0.000000   
...                    ...    ...   ...    ...       ...      ...       ...   
534        Aaron Cresswell     20  1586      0  0.419239        1  0.937578   
534        Aaron Cresswell     11   990      0  0.264341        3  1.606536   
534        Aaron Cresswell     31  2739      3  1.032538        0  3.006657   
7991        Aaron Connolly     24  1279      3  4.553526        1  0.562017   
7991        Aaron Connolly      8   425      1  1.553084        1  0.132376   

           shots  key_passes  yellow_cards  red_cards position      team_name  \
player_id                                                                       
250            0           0             0          0       GK     Ingolstadt   
250            0           0             0          0     GK S     Ingolstadt   
250            0           0             0          0     GK S    Aston Villa   
2646           0           0             0          0        M  Werder Bremen   
2646           1           0             0          0        S  Werder Bremen   
...          ...         ...           ...        ...      ...            ...   
534           11          17             1          0      D S       West Ham   
534            4          20             0          0        D       West Ham   
534           18          26             7          0        D       West Ham   
7991          36           6             0          0      F S       Brighton   
7991           9           4             0          0      F S       Brighton   

           npg      npxG   xGChain  xGBuildup  year  
player_id                                            
250          0  0.000000  0.167960   0.167960  2015  
250          0  0.000000  0.593702   0.593702  2016  
250          0  0.000000  0.194507   0.194507  2019  
2646         0  0.000000  0.013800   0.013800  2014  
2646         0  0.083136  0.083136   0.000000  2015  
...        ...       ...       ...        ...   ...  
534          0  0.419239  2.039328   1.581602  2018  
534          0  0.264341  3.518969   2.893820  2020  
534          3  1.032538  6.834942   4.891464  2019  
7991         3  4.553526  4.356782   0.354199  2019  
7991         1  1.553084  1.683810   0.031006  2020  

[18633 rows x 18 columns]

In [14]:
print(df.sort_values(['team_name', 'goals' ], ascending=[True, False]).head(3))

            player_name  games  time  goals         xG  assists        xA  \
player_id                                                                   
1125       Carlos Bacca     38  3179     18  14.870253        2  3.657710   
2014       Jérémy Menez     33  2713     16  10.936869        4  4.379692   
1125       Carlos Bacca     32  2152     13  14.850961        3  2.608146   

           shots  key_passes  yellow_cards  red_cards position team_name  npg  \
player_id                                                                       
1125          77          41             2          0      F S  AC Milan   16   
2014          78          46             3          1    F M S  AC Milan    8   
1125          56          19             3          0      F S  AC Milan    9   

                npxG    xGChain  xGBuildup  year  
player_id                                         
1125       13.347656  18.540670   3.277732  2015  
2014        4.846598  13.755916   6.889687  2014  
1125

## Subconjuntos y filtros de datos

In [15]:
df['team_name'].unique()

<StringArray>
[          'Aston Villa',               'Everton',           'Southampton',
             'Leicester',  'West Bromwich Albion',        'Crystal Palace',
               'Chelsea',              'West Ham',             'Tottenham',
               'Arsenal',
 ...
            'Ingolstadt',             'Darmstadt',                 'Carpi',
           'GFC Ajaccio',   'Queens Park Rangers',               'Cordoba',
               'Almeria',                'Cesena',                 'Parma',
 'Evian Thonon Gaillard']
Length: 147, dtype: str

### Filtrar filas

El método `.isin()` en Pandas sirve para filtrar y seleccionar datos de un DataFrame o Series, comprobando de manera eficiente si los elementos están contenidos en una lista.

A continuación se muestran diferentes formas de filtrado:


In [16]:
df[df.player_name == 'Aaron Connolly']

player_name  games  time  goals        xG  assists        xA  \
player_id                                                                    
7991       Aaron Connolly      8   425      1  1.553084        1  0.132376   
7991       Aaron Connolly     24  1279      3  4.553526        1  0.562017   

           shots  key_passes  yellow_cards  red_cards position team_name  npg  \
player_id                                                                       
7991           9           4             0          0      F S  Brighton    1   
7991          36           6             0          0      F S  Brighton    3   

               npxG   xGChain  xGBuildup  year  
player_id                                       
7991       1.553084  1.683810   0.031006  2020  
7991       4.553526  4.356782   0.354199  2019

In [17]:
df[df['player_name'] == 'Aaron Connolly']

player_name  games  time  goals        xG  assists        xA  \
player_id                                                                    
7991       Aaron Connolly      8   425      1  1.553084        1  0.132376   
7991       Aaron Connolly     24  1279      3  4.553526        1  0.562017   

           shots  key_passes  yellow_cards  red_cards position team_name  npg  \
player_id                                                                       
7991           9           4             0          0      F S  Brighton    1   
7991          36           6             0          0      F S  Brighton    3   

               npxG   xGChain  xGBuildup  year  
player_id                                       
7991       1.553084  1.683810   0.031006  2020  
7991       4.553526  4.356782   0.354199  2019

In [18]:
df.query('player_name == "Aaron Connolly"')

player_name  games  time  goals        xG  assists        xA  \
player_id                                                                    
7991       Aaron Connolly      8   425      1  1.553084        1  0.132376   
7991       Aaron Connolly     24  1279      3  4.553526        1  0.562017   

           shots  key_passes  yellow_cards  red_cards position team_name  npg  \
player_id                                                                       
7991           9           4             0          0      F S  Brighton    1   
7991          36           6             0          0      F S  Brighton    3   

               npxG   xGChain  xGBuildup  year  
player_id                                       
7991       1.553084  1.683810   0.031006  2020  
7991       4.553526  4.356782   0.354199  2019

### Uniendo lo aprendido

In [19]:
# Goles de Cristiano Ronaldo realizó en 2014
cr7 = df[(df.player_name == 'Cristiano Ronaldo') & (df.year == 2014)]
cols = ['player_name', 'year', 'goals' ]
print(cr7[cols])

                 player_name  year  goals
player_id                                
2371       Cristiano Ronaldo  2014     48


In [20]:
# Goles de Ronaldo, Messi y Suárez en 2015
players = ['Cristiano Ronaldo', 'Lionel Messi', 'Luis Suárez']
top_players = df[(df.player_name.isin(players)) & (df.year == 2015)]
print(top_players[cols])

                 player_name  year  goals
player_id                                
2098             Luis Suárez  2015     40
2097            Lionel Messi  2015     26
2371       Cristiano Ronaldo  2015     35


### Subsetting con `.loc`


`.loc` es un indexador fundamental utilizado para seleccionar, filtrar y manipular subconjuntos de datos en un DataFrame o Series basándose en etiquetas (nombres) de filas y columnas, en lugar de sus posiciones numéricas.

In [21]:
# Goles de Ronaldo, Messi y Suárez en 2015
df = df.reset_index() #convierte el índice actual en una columna
df = df.set_index('player_name') #Define 'player_name' como índice del DF
players = ['Cristiano Ronaldo', 'Lionel Messi', 'Luis Suárez']
cols = [ 'year', 'goals' ]
top_players = df.loc[players, cols]
print(top_players.query('year == 2015')) #query() filtra filas usando SQL

                   year  goals
player_name                   
Cristiano Ronaldo  2015     35
Lionel Messi       2015     26
Luis Suárez        2015     40


## Aplicar funciones a un DataFrame

- `.sum()` → suma todos los valores de una columna.
- `.mean()` → promedio de los valores de una columna.
- `.max()` → valor máximo de una columna.
- `.min()` → valor mínimo de una columna.
- `.cumsum()` → en cada fila va poniendo la suma acumulada de una columna.
- `.cummax()` → en cada fila va poniendo el valor máximo encontrado en orden de una columna.
- `.value_counts(sort=True)` → cuenta los distintos valores que existen en una columna (muy útil para contar categorías) y los ordena descendientemente.
- `.drop_duplicates(subset='col_name')` → elimina duplicados en una columna.

También es posible pasar funciones que nosotros mismos hayamos creado u otras funciones importadas de otras librerías. Para ello se usa `.agg(function)`.

In [22]:
# Máxima cantidad de goles hechas por un jugador en una temporada
max_goles = df[ 'goals'].max()
cols = ['team_name','year', 'goals']
print(df[df.goals == max_goles][cols])

                     team_name  year  goals
player_name                                
Cristiano Ronaldo  Real Madrid  2014     48


In [23]:
# Porcentaje de disparos al arco que terminan en gol
total_shots = df[ 'shots'].sum()
total_goals = df[ 'goals'].sum()
print(round(total_goals / total_shots * 100, 2))

10.6


## Agrupar datos

- `df.groupby()`

In [26]:
# Top 10 jugadores que más goles han marcado
top_players = df.groupby('player_name')['goals'].sum()
print(top_players.sort_values( ascending=False).head(10))

player_name
Lionel Messi                 205
Cristiano Ronaldo            194
Robert Lewandowski           174
Luis Suárez                  154
Harry Kane                   148
Pierre-Emerick Aubameyang    141
Sergio Agüero                128
Edinson Cavani               125
Ciro Immobile                119
Mohamed Salah                117
Name: goals, dtype: int64


In [28]:
# Top 5 equipos con más tarjetas rojas y amarillas
red_card = df.groupby('team_name')[['red_cards', 'yellow_cards']].sum()
print(red_card.sort_values( 'red_cards',ascending=False).head())

           red_cards  yellow_cards
team_name                         
AC Milan          43           513
Genoa             42           562
Valencia          39           598
Bologna           38           477
Monaco            35           431


## Crear columnas


In [29]:
df['goals_assists'] = df.goals + df.assists
best_players = df.groupby('player_name')['goals_assists'].sum()
print(best_players.sort_values(ascending=False).head())

player_name
Lionel Messi          293
Cristiano Ronaldo     246
Luis Suárez           224
Robert Lewandowski    203
Harry Kane            178
Name: goals_assists, dtype: int64


## Ejercicios

## 1. Usa `df.describe()` y cuéntanos todos los insights que te dio y qué podría significar cada uno.

In [30]:
print(df.describe())

          player_id         games          time         goals            xG  \
count  18633.000000  18633.000000  18633.000000  18633.000000  18633.000000   
mean    3107.432405     17.072291   1207.483282      1.624054      1.660439   
std     2439.349524     11.375527    975.753086      3.252418      2.974699   
min        1.000000      1.000000      1.000000      0.000000      0.000000   
25%     1118.000000      7.000000    326.000000      0.000000      0.060762   
50%     2337.000000     16.000000    976.000000      0.000000      0.560141   
75%     5074.000000     27.000000   2004.000000      2.000000      1.900353   
max     9177.000000     38.000000   3420.000000     48.000000     39.308761   

            assists            xA         shots    key_passes  yellow_cards  \
count  18633.000000  18633.000000  18633.000000  18633.000000  18633.000000   
mean       1.138088      1.165425     15.323888     11.354801      2.453979   
std        1.936238      1.735366     20.934645    

Para cada variable numérica `.describe()` da:
- **count**: Cuántos registros válidos hay
- **mean**: Promedio
- **std**: dispersión (qué tan desiguales son los datos)
- **min / max**: Extremos
- **25%, 50%, 75%** Cuartiles (distribución)

Como dato inicial, contamos con 18,633 registros en el dataset

### Participación y minutos

`games` y `time`
- Promedio de partido: `17`
- Promedio de minutos: `1207 (Aprox 13 partidos completos)`
- Máximo: `3420 minutos temporada completa (38 partidos)`

**📌 Insight**
1. La mayoria de jugadores no juegan una temporada completa
2. Hay mucha rotación, supletes y jugadores con pocos minutos.
3. El 25% de los jugadores juega menos de 326 minutos

### Asistencias y creación de juego

`assists` y `xA`
- Mediana de asistencias: `0`
- Media xA: ` 1.16`
- Máximo xA: ` 20.6`

**📌 Insight**
1. Crear goles es aún más raro que anotarlos
2. Muy pocos jugadores concentran la creación ofensiva

`key_passes`
- Mediana: `5`
- Máximo: `146`

**📌 Insight**
1. Algunos jugadores son cerebros del equipo
2. Desigualdad entre roles ofensivos

Conclusión: Dataset muy sesgado (pocos crack, muchos roles secundarios)



## 2. Encuentra la cantidad de goles totales hechos por cada posición.

In [34]:
total_goals = df.groupby('position')['goals'].sum()
print(total_goals.sort_values(ascending=False))

position
F M S      9718
F S        8118
M S        4536
D S        1494
D          1238
D M S      1213
F          1022
F M         982
M           764
D F M S     637
D M         360
S            90
D F M        62
D F S        25
GK            1
GK S          1
D F           0
Name: goals, dtype: int64


## 3. ¿Cuál es el jugador que más tiempo ha jugado y cuál es la media de tiempo de todos los jugadores? Esta pregunta tiene truco porque hay muchos jugadores suplentes que no juegan casi nunca y van a sesgar los datos, ¿cómo podrías solucionar esto?

In [39]:
# 1) Tiempo total jugado por cada jugador (agregamos por jugador)
time_per_player = df.groupby('player_name')['time'].sum()

# 2) Jugador que MÁS tiempo ha jugado
player_most_time = time_per_player.idxmax()
max_time = time_per_player.max()

print(f"Jugador con más tiempo jugado: {player_most_time} ({max_time} minutos)")

# 3) Media de tiempo jugado por jugador (sin sesgo por temporadas)
mean_time = time_per_player.mean()
print(f"Media de tiempo jugado por jugador: {mean_time:.2f} minutos")

# 4) Corrección del sesgo: filtrar jugadores con participación real
regular_players = time_per_player[time_per_player >= 900]  # ~10 partidos

mean_time_regulars = regular_players.mean()
print(f"Media de tiempo (jugadores regulares): {mean_time_regulars:.2f} minutos")



Jugador con más tiempo jugado: Danilo (31012 minutos)
Media de tiempo jugado por jugador: 3645.93 minutos
Media de tiempo (jugadores regulares): 5595.91 minutos


## 4. ¿Qué posición es la que más tiempo juega? ¿Y la que menos?

In [43]:
# 1) Tiempo total jugado por cada posición
time_position = df.groupby('position')['time'].sum()

# 2) Jugador que MÁS tiempo ha jugado
position_most_time = time_position.idxmax()
max_time_p = time_position.max()

# 3) Posición con MENOS tiempo jugado
position_less_time = time_position.idxmin()
min_time_p = time_position.min()



print(
    f"Posición con más tiempo jugado: {position_most_time} ({max_time_p} minutos)\n"
    f"Posición con menos tiempo jugado: {position_less_time} ({min_time_p} minutos)"
)

Posición con más tiempo jugado: M S (4296344 minutos)
Posición con menos tiempo jugado: D F (885 minutos)


## 5. ¿Qué porcentaje de goles son hechos por penales? La columna npg significa: goles hechos SIN penales.

In [44]:
# 1) Goles totales
total_goals = df['goals'].sum()

# 2) Goles sin penales
non_penalty_goals = df['npg'].sum()

# 3) Goles de penal
penalty_goals = total_goals - non_penalty_goals

# 4) Porcentaje de goles de penal
penalty_percentage = (penalty_goals / total_goals) * 100

print(f"El {penalty_percentage:.2f}% de los goles fueron de penal")


El 8.97% de los goles fueron de penal


## 6. ¿Qué porcentaje de goles son hechos con asistencias?

In [45]:
# 1) Goles totales
total_goals = df['goals'].sum()

# 2) Goles con asistencia (aproximado por total de asistencias)
assisted_goals = df['assists'].sum()

# 3) Porcentaje de goles con asistencia
assisted_percentage = (assisted_goals / total_goals) * 100

print(f"El {assisted_percentage:.2f}% de los goles fueron con asistencia")


El 70.08% de los goles fueron con asistencia


## 7. ¿Cuál fue el equipo que más goles hizo cada año? ¿Salió campeón ese año?

In [46]:
# 1) Goles totales por equipo y año
goals_team_year = df.groupby(['year', 'team_name'])['goals'].sum().reset_index()

# 2) Equipo con más goles por año
top_scorers = goals_team_year.loc[
    goals_team_year.groupby('year')['goals'].idxmax()
]

print(top_scorers)


     year            team_name  goals
71   2014          Real Madrid    116
106  2015            Barcelona    109
204  2016            Barcelona    114
351  2017      Manchester City    103
458  2018  Paris Saint Germain    103
542  2019      Manchester City    100
600  2020        Bayern Munich     34


## 8. ¿Cuántos jugadores jugaron todos los años del dataset? ¿Qué porcentaje del total representan?

In [48]:
# Total de años en el dataset
total_years = df['year'].nunique()

# Años jugados por cada jugador
df = df.reset_index()
years_per_player = df.groupby('player_name')['year'].nunique()

# Jugadores que jugaron todos los años
players_all_years = years_per_player[years_per_player == total_years]

# Cantidad de jugadores
num_players_all_years = players_all_years.count()

# Total de jugadores
total_players = df['player_name'].nunique()

# Porcentaje
percentage = (num_players_all_years / total_players) * 100

print(f"Jugadores que jugaron todos los años: {num_players_all_years}")
print(f"Porcentaje sobre el total: {percentage:.2f}%")


Jugadores que jugaron todos los años: 540
Porcentaje sobre el total: 8.75%
